src/lightning_lite/utilities/device_parser.py

```diff
@@ -286,16 +287,30 @@ def _parse_tpu_cores_str(tpu_cores: str) -> Union[int, List[int]]:
     return [int(x.strip()) for x in tpu_cores.split(",") if len(x) > 0]
 
 
+@lru_cache
 def num_cuda_devices() -> int:
-    """Returns the number of GPUs available.
+    """Returns the number of available CUDA devices.
 
     Unlike :func:`torch.cuda.device_count`, this function does its best not to create a CUDA context for fork support,
     if the platform allows it.
     """
-    if "fork" not in torch.multiprocessing.get_all_start_methods() or _is_forking_disabled():
+
+    # Code adapted from Nikita Shulga, @malfet:
+    # https://github.com/pytorch/pytorch/issues/83973#issuecomment-1238633755
+    try:
+        nvml_h = CDLL("libnvidia-ml.so.1")
+    except OSError:
+        # Can't load NVML shared libraries, fall back to torch.cuda
         return torch.cuda.device_count()
-    with multiprocessing.get_context("fork").Pool(1) as pool:
-        return pool.apply(torch.cuda.device_count)
+
+    rc = nvml_h.nvmlInit()
+    assert rc == 0
```

Comment:
I would switch these assertions for gotos into exception handling code that falls back to something sensible, `device_count` or the previous `get_context` solution